solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java

```diff
@@ -553,6 +560,82 @@ private void setQuery(ResponseBuilder rb, Elevation elevation) {
     }
   }
 
+  /**
+   * Updates any filters that have been tagged for exclusion. Filters can be tagged for exclusion
+   * via the syntax fq={!tag=t1}field1:value1&elevate.excludeTags=t1 This method modifies each
+   * "excluded" filter so that it becomes a boolean OR of the original filter with an "include
+   * query" that matches the elevated documents by their IDs. To be clear, the "excluded" filters
+   * are not ignored entirely, but rather broadened so that the elevated documents are allowed
+   * through.
+   */
+  private void setFilters(ResponseBuilder rb, Elevation elevation) {
+    SolrParams params = rb.req.getParams();
+
+    String tagStr = params.get(QueryElevationParams.ELEVATE_EXCLUDE_TAGS);
+    if (StringUtils.isEmpty(tagStr)) {
+      // the parameter that specifies tags for exclusion was not provided or had no value
+      return;
+    }
+
+    List<String> excludeTags = StrUtils.splitSmart(tagStr, ',');
+    excludeTags.removeIf(s -> StringUtils.isBlank(s));
+    if (excludeTags.isEmpty()) {
+      // the parameter that specifies tags for exclusion was provided but the tag names were blank
+      return;
+    }
+
+    List<Query> filters = rb.getFilters();
+    if (filters == null || filters.isEmpty()) {
+      // no filters were provided
+      return;
+    }
+
+    Map<?, ?> tagMap = (Map<?, ?>) rb.req.getContext().get("tags");
+    if (tagMap == null) {
+      // no filters were tagged
+      return;
+    }
+
+    // TODO: this code is copied from FacetProcessor#handleFilterExclusions()
+    // duplication could be avoided by placing this code in a common utility method, perhaps in
+    // QueryUtils
```

Comment:
As a first step, I made a private static utility method in QueryElevationComponent to house this logic. Seeing that the signature is:

private static Set<Query> getTaggedQueries(ResponseBuilder rb, List<String> desiredTags)

...I'm wondering if the method could ultimately go on ResponseBuilder itself, where it could be:

public Set<Query> getTaggedQueries(List<String> tags)

I'll continue thinking re: the TaggedQueryContext idea you described... but wondering if you had any feedback on this possibility?

Reply:
Maybe `ResponseBuilder`.  There is no great place because we don't capture this feature into any one class.  RB is mostly a POJO so this method would seem quite random there.  But not bad.  I support that.

Reply:
FYI I'm planning to leave this change till the other items in the PR are resolved. And I have a little bit of reluctance to broadening this PR to include changes outside QueryElevationComponent itself :) But if you'd like to see how this refactor would look right now, LMK and I'll go for it.

Reply:
Let's do it; okay?  ResponseBuilder is fine.

In my experience, better to do the refactoring as part of an issue like this because it will likely be forgotten if left to the future.

Reply:
Yes, thx for the encouragement to do it now :) Two questions came up in implementing the `ResponseBuilder` approach. 1) I noticed that `FacetProcessor#handleFilterExclusions()` has a comment about wanting to _remove_ the `ResponseBuilder` dependency there -- how significant is this? 2) I noticed that the facet processor code gets the "tagMap" via the `SolrQueryRequest` it finds in its `FacetContext`. Can I assume the `SolrQueryRequest` inside the `ResponseBuilder` (obtained via `SolrRequestInfo.getRequestInfo().getResponseBuilder()`) would be the same as the one in the FacetContext, or at least that its request context would have the same tags inside?  To make things clearer, I sketched out what the refactor would look like with RB as the destination, as well as two other possible destinations: QueryUtils and SolrQueryRequest itself.  Let me know if RB still seems best to you or if either of the others looks reasonable. See: https://github.com/rseitz/solr/commit/a0675c51eed9f29b35d0a1a2fe626dcb16812077

Reply:
>  Can I assume the SolrQueryRequest inside the ResponseBuilder (obtained via SolrRequestInfo.getRequestInfo().getResponseBuilder()) would be the same as the one in the FacetContext, or at least that its request context would have the same tags inside?

Yes.  Tests would fail if we're wrong on this, I'd think.

Heh, funny, that TODO to not depend on ResponseBuilder was added by me.  It wasn't an important TODO.

I changed my mind on QueryUtils suitability.  First I thought "no" -- all methods on QueryUtils (from my memory) are about looking at Query classes; not about SolrQueryRequest.  But I see a very recent refactoring where someone added another method here, parseFilterQueries that takes the SQR.  I suppose why not, let this class increase in scope.  Besides... "ResponseBuilder" does not at all sound like the name of a class that would have methods pertaining to accessing queries in a *request*.

Reply:
OK, I've done the deed. Agree, QueryUtils feels maybe a little better than RB given the presence of parseFilterQueries there. Thx again for the feedback and let me know how it looks.